nixos/tests/rosenpass/default.nix

```diff
@@ -0,0 +1,193 @@
+import ../make-test-python.nix ({ ... }:
+let
+  deviceName = "rp0";
+
+  server = {
+    ip = "fe80::1";
+    wg = {
+      public = "mQufmDFeQQuU/fIaB2hHgluhjjm1ypK4hJr1cW3WqAw=";
+      secret = "4N5Y1dldqrpsbaEiY8O0XBUGUFf8vkvtBtm8AoOX7Eo=";
+      listen = 10000;
+    };
+    rp = {
+      secret = ./server-pqsk;
+      public = ./server-pqpk;
+    };
+  };
+  client = {
+    ip = "fe80::2";
+    wg = {
+      public = "Mb3GOlT7oS+F3JntVKiaD7SpHxLxNdtEmWz/9FMnRFU=";
+      secret = "uC5dfGMv7Oxf5UDfdPkj6rZiRZT2dRWp5x8IQxrNcUE=";
+    };
+    rp = {
+      secret = ./client-pqsk;
+      public = ./client-pqpk;
+    };
+  };
+in
+{
+  name = "rosenpass";
+
+  nodes =
+    let
+      etcPath = config: name: "/etc/" + config.environment.etc.${name}.target;
+      etcPathBin = config: name: (etcPath config name) + ".bin";
+
+      shared =
+        peer:
+        { config
+        , modulesPath
+        , ...
+        }: {
+          imports = [ "${modulesPath}/services/networking/rosenpass.nix" ];
+
+          boot.kernelModules = [ "wireguard" ];
+
+          services.rosenpass = {
+            enable = true;
+            publicKeyFile = etcPathBin config "rosenpass/pqpk";
+            secretKeyFile = etcPathBin config "rosenpass/pqsk";
+            defaultDevice = deviceName;
+          };
+
+          networking.firewall.allowedUDPPorts = [ 9999 ];
+
+          systemd = {
+            network = {
+              enable = true;
+              networks."rosenpass" = {
+                matchConfig.Name = deviceName;
+                networkConfig.IPForward = true;
+                address = [ "${peer.ip}/64" ];
+              };
+
+              netdevs."10-rp0" = {
+                netdevConfig = {
+                  Kind = "wireguard";
+                  Name = deviceName;
+                };
+                wireguardConfig.PrivateKeyFile = etcPath config "wireguard/wgsk";
+              };
+            };
+
+            # Rosenpass requires keys in binary format.
+            # However, we want to avoid storing binary files in the nixpkgs repository.
+            # We are providing keys in base64 encoded format
+            # in `/etc/rosenpass/**/*pq{p,s}k` via `environment.etc`.
+            # This service decodes those files back into the binary format that
+            # Rosenpass expects.
+            # See also `etcPath x y` and `etcPathBin x y`.
```

Comment:
Would it be possible to generate the keys as part of the test?

Reply:
Yes, but it would be slightly more involved. The public parts would have to be shared between the two VMs at runtime.